src/main/python/org/broadinstitute/hellbender/gcnvkernel/postprocess/viterbi_segmentation.py

```diff
@@ -0,0 +1,371 @@
+            log_q_tau_tk = self.log_q_tau_tk[contig_interval_indices, :]
+
+            # copy-number log emission probability for contig intervals
+            copy_number_log_emission_contig_tc = copy_number_log_emission_tc[contig_interval_indices, :]
+
+            # get HMM specs
+            hmm_specs = self.get_copy_number_hmm_specs(
+                pi_jkc, self.cnv_stay_prob_t_j[contig], log_q_tau_tk, t_to_j_map)
+            log_prior_c = hmm_specs[0]
+            log_trans_contig_tcc = hmm_specs[1]
+
+            # run forward-back algorithm
+            fb_result = self.theano_forward_backward.perform_forward_backward_no_admix(
+                log_prior_c, log_trans_contig_tcc, copy_number_log_emission_contig_tc)
+            log_posterior_prob_tc = fb_result[0]
+            log_data_likelihood = fb_result[1]
+            alpha_tc = fb_result[2]
+            beta_tc = fb_result[3]
+
+            # run viterbi algorithm
+            viterbi_path_t_contig = self.theano_viterbi.get_viterbi_path(
+                log_prior_c, log_trans_contig_tcc, copy_number_log_emission_contig_tc)
+
+            # initialize the segment quality calculator
+            segment_quality_calculator = HMMSegmentationQualityCalculator(
+                copy_number_log_emission_contig_tc, log_trans_contig_tcc,
+                alpha_tc, beta_tc, log_posterior_prob_tc, log_data_likelihood)
+
+            # coalesce into piecewise constant copy-number segments, calculate qualities
+            for copy_number_call, start_index, end_index in self.coalesce_seq_into_segments(viterbi_path_t_contig):
+                num_spanning_intervals = end_index - start_index + 1
+                segment = IntegerCopyNumberSegment(contig,
+                                                   contig_interval_list[start_index].start,
+                                                   contig_interval_list[end_index].end,
+                                                   num_spanning_intervals,
+                                                   copy_number_call)
+                if num_spanning_intervals > 1:
+                    segment.some_quality = segment_quality_calculator.get_segment_some_quality(
+                        start_index, end_index, copy_number_call)
+                    segment.exact_quality = segment_quality_calculator.get_segment_exact_quality(
+                        start_index, end_index, copy_number_call)
+                    segment.start_quality = segment_quality_calculator.get_segment_start_quality_direct(
+                        start_index, copy_number_call)
+                    segment.end_quality = segment_quality_calculator.get_segment_end_quality_direct(
+                        end_index, copy_number_call)
+
+                    # exact quality calculation is prone to numerical instability and may overflow
+                    # in that case, some quality will be reported
+                    if np.isinf(segment.exact_quality):
```

Comment:
This seems non-ideal.  Does this only occur for high quality segments, long segments, etc.?  What causes the overflow?

Reply:
I thought quite a bit about this but I can't find an easy solution. It only occurs for HQ segments where `log P(all intervals = call)` ~ round-off error such that `log P(some intervals != call)` is ~ -inf and unreliable. Unfortunately, it is not easy to calculate the latter directly. The same thing occurred for breakpoint quality calculation for HQ segments and that's why I wrote those additional `_direct` methods. In that case, my workaround was to calculate the complementary log P directly by summing over complementary events.

Reply:
Perhaps we could neglect correlations while calculating the exact quality, i.e. `log P(all intervals = call) ~ \sum_j log P (interval_j = call)`. This is likely to be less prone to round-off error.

Reply:
I implemented an adaptive method that switches between a robust upper bound approximation and the (shaky) exact calculation in cases the exact calculation is expected to be robust, too.